sysbuild: MCUboot with ED25519 and KMU via PSA support

de-nordic opened the pull request:
The commit will enforce building nrf54l15 with PSA enabled
ED25519, with CONFIG_NRF_SECURITY=y.
The commit adds SB_CONFIG_MCUBOOT_SIGNATURE_USING_KMU that allows
to build MCUboot for signature verification via KMU instead of
compiled in keys.

No longer sdk-zephyr changes.

Building with:
```
west build  -p -d builds/hello_54_kmu -b nrf54l15dk/nrf54l15/cpuapp zephyr/samples/hello_world/ -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DSB_CONFIG_BOOT_SIGNATURE_TYPE_ED25519=y  -Dmcuboot_CONFIG_PM_PARTITION_SIZE_MCUBOOT=0x10000  -DSB_CONFIG_MCUBOOT_SIGNATURE_USING_KMU=y
```

gives KMU build. Additionally option `-DSB_CONFIG_BOOT_SIGNATURE_KEY_FILE=<>` may be used to select signature key file.

NordicBuilder commented:
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| mcuboot | https://github.com/nrfconnect/sdk-mcuboot/commit/5f95fecfa27e5b9924a91c69bca023e605fe751e | https://github.com/nrfconnect/sdk-mcuboot/pull/341 | [nrfconnect/sdk-mcuboot#341/files](https://github.com/nrfconnect/sdk-mcuboot/pull/341/files) |
| zephyr | https://github.com/nrfconnect/sdk-zephyr/commit/5d10d27a0a978816153764e8f7024cbbc3fae197 (main) | https://github.com/nrfconnect/sdk-zephyr/pull/2056 | [nrfconnect/sdk-zephyr#2056/files](https://github.com/nrfconnect/sdk-zephyr/pull/2056/files) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-17584
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 1
### Inputs:
#### Sources:
sdk-nrf: PR head: [af70474c105412fb21aeef94baaa0b5d51b7c9af](https://github.com/nrfconnect/sdk-nrf/pull/17584/commits/af70474c105412fb21aeef94baaa0b5d51b7c9af)
mcuboot: PR head: [c4b41bfb5d849f42b44956c54e00fd036251ebcb](https://github.com/nrfconnect/sdk-mcuboot/pull/341/commits/c4b41bfb5d849f42b44956c54e00fd036251ebcb)
zephyr: PR head: [5d7765b58beab14bba5abb4b43b5c18f44057cf5](https://github.com/nrfconnect/sdk-zephyr/pull/2056/commits/5d7765b58beab14bba5abb4b43b5c18f44057cf5)
<details><summary> more details</summary>

sdk-nrf:
> PR head: [af70474c105412fb21aeef94baaa0b5d51b7c9af](https://github.com/nrfconnect/sdk-nrf/pull/17584/commits/af70474c105412fb21aeef94baaa0b5d51b7c9af)
> merge base: [7b1f88d8009719d50fd389487257675246ff4a70](https://github.com/nrfconnect/sdk-nrf/commit/7b1f88d8009719d50fd389487257675246ff4a70)
> target head (main): [7b1f88d8009719d50fd389487257675246ff4a70](https://github.com/nrfconnect/sdk-nrf/commit/7b1f88d8009719d50fd389487257675246ff4a70)
> [Diff](https://github.com/nrfconnect/sdk-nrf/compare/7b1f88d8009719d50fd389487257675246ff4a70..af70474c105412fb21aeef94baaa0b5d51b7c9af)

mcuboot:
> PR head: [c4b41bfb5d849f42b44956c54e00fd036251ebcb](https://github.com/nrfconnect/sdk-mcuboot/pull/341/commits/c4b41bfb5d849f42b44956c54e00fd036251ebcb)
> merge base: [5f95fecfa27e5b9924a91c69bca023e605fe751e](https://github.com/nrfconnect/sdk-mcuboot/commit/5f95fecfa27e5b9924a91c69bca023e605fe751e)
> target head (main): [de524e93b604d54541736875769c851ed6a3ed3e](https://github.com/nrfconnect/sdk-mcuboot/commit/de524e93b604d54541736875769c851ed6a3ed3e)
> [Diff](https://github.com/nrfconnect/sdk-mcuboot/compare/5f95fecfa27e5b9924a91c69bca023e605fe751e..c4b41bfb5d849f42b44956c54e00fd036251ebcb)

zephyr:
> PR head: [5d7765b58beab14bba5abb4b43b5c18f44057cf5](https://github.com/nrfconnect/sdk-zephyr/pull/2056/commits/5d7765b58beab14bba5abb4b43b5c18f44057cf5)
> merge base: [4322f5169f4c0f24e3712de061a2fec5a29098a7](https://github.com/nrfconnect/sdk-zephyr/commit/4322f5169f4c0f24e3712de061a2fec5a29098a7)
> target head (main): [5d10d27a0a978816153764e8f7024cbbc3fae197](https://github.com/nrfconnect/sdk-zephyr/commit/5d10d27a0a978816153764e8f7024cbbc3fae197)
> [Diff](https://github.com/nrfconnect/sdk-zephyr/compare/4322f5169f4c0f24e3712de061a2fec5a29098a7..5d7765b58beab14bba5abb4b43b5c18f44057cf5)
</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (4)</summary>

```
bootloader
│  ├── mcuboot
│  │  ├── boot
│  │  │  ├── zephyr
│  │  │  │  ├── CMakeLists.txt
│  │  │  │  │ Kconfig
west.yml
zephyr
│  ├── share
│  │  ├── sysbuild
│  │  │  ├── images
│  │  │  │  ├── bootloader
│  │  │  │  │  │ Kconfig
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :orange_circle: Toolchain
- :orange_circle: Build twister
- :orange_circle: Integration tests
   - :orange_circle: test-fw-nrfconnect-boot
   - :orange_circle: test-fw-nrfconnect-rs
   - :orange_circle: test-fw-nrfconnect-fem
   - :orange_circle: test-sdk-mcuboot

<details><summary>Disabled integration tests</summary>

- 
   - desktop52_verification
   - doc-internal
   - test_ble_nrf_config
   - test-fw-nrfconnect-apps
   - test-fw-nrfconnect-ble_mesh
   - test-fw-nrfconnect-ble_samples
   - test-fw-nrfconnect-chip
   - test-fw-nrfconnect-nfc
   - test-fw-nrfconnect-nrf-iot_cloud
   - test-fw-nrfconnect-nrf-iot_libmodem-nrf
   - test-fw-nrfconnect-nrf-iot_lwm2m
   - test-fw-nrfconnect-nrf-iot_mosh
   - test-fw-nrfconnect-nrf-iot_nrf_provisioning
   - test-fw-nrfconnect-nrf-iot_positioning
   - test-fw-nrfconnect-nrf-iot_samples
   - test-fw-nrfconnect-nrf-iot_serial_lte_modem
   - test-fw-nrfconnect-nrf-iot_thingy91
   - test-fw-nrfconnect-nrf-iot_zephyr_lwm2m
   - test-fw-nrfconnect-nrf_crypto
   - test-fw-nrfconnect-proprietary_esb
   - test-fw-nrfconnect-ps
   - test-fw-nrfconnect-rpc
   - test-fw-nrfconnect-tfm
   - test-fw-nrfconnect-thread
   - test-fw-nrfconnect-zigbee
   - test-low-level
   - test-sdk-audio
   - test-sdk-dfu
   - test-sdk-find-my
   - test-sdk-pmic-samples
   - test-sdk-sidewalk
   - test-sdk-wifi
   - test-secdom-samples-public
</details>



*Note: This message is automatically posted and updated by the CI*

de-nordic committed:
sysbuild: MCUboot with ED25519 and KMU via PSA support

The commit will enforce building nrf54l15 with PSA enabled
ED25519, with CONFIG_NRF_SECURITY=y.
The commit adds SB_CONFIG_MCUBOOT_SIGNATURE_USING_KMU that allows
to build MCUboot for signature verification via KMU instead of
compiled in keys.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on sysbuild/Kconfig.mcuboot:
```diff
@@ -118,6 +118,18 @@ choice BOOT_SIGNATURE_TYPE
+	default MCUBOOT_CRYPTO_LIBRARY_DEFAULT
+
+config MCUBOOT_CRYPTO_LIBRARY_PSA
+	bool "Enforce use of PSA as cryptographic library"
```
`depends on BOOT_SIGNATURE_TYPE_ED25519`

nordicjm commented on sysbuild/CMakeLists.txt:
```diff
@@ -224,6 +224,17 @@ function(${SYSBUILD_CURRENT_MODULE_NAME}_pre_cmake)
     endforeach()
 
+    if(SB_CONFIG_MCUBOOT_CRYPTO_LIBRARY_PSA)
+      if (SB_CONFIG_BOOT_SIGNATURE_TYPE_ED25519)
+        set_config_bool(mcuboot CONFIG_NRF_SECURITY y)
+        set_config_bool(mcuboot CONFIG_MBEDTLS n)
+        set_config_bool(mcuboot CONFIG_BOOT_ED25519_PSA y)
+        set_config_string(mcuboot CONFIG_MBEDTLS_CFG_FILE "nrf-config.h")
+      else()
+        message(ERROR "Switching to PSA is currently available only for ED25519")
+      endif()
```
```suggestion
      set_config_bool(mcuboot CONFIG_NRF_SECURITY y)
      set_config_bool(mcuboot CONFIG_MBEDTLS n)
      set_config_bool(mcuboot CONFIG_BOOT_ED25519_PSA y)
      set_config_string(mcuboot CONFIG_MBEDTLS_CFG_FILE "nrf-config.h")
```

nvlsianpu commented:
@de-nordic the manifest commit can be dropped. The reference is already ahead of that.

NordicBuilder commented:
### Memory footprint analysis revealed the following potential issues

***applications.nrf_desktop.zrelease[nrf54h20dk/nrf54h20/cpuapp]***: ***RAM*** size  ***increased by 948[B]*** in comparison to the ***main[b02dd1b4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=applications&var-submodule=nrf_desktop&var-scenario=zrelease&var-testcase=none&var-arch=arm&var-platform=nrf54h20dk/nrf54h20/cpuapp) (cc: @MarekPieta)
***applications.nrf_desktop.zrelease[nrf54h20dk/nrf54h20/cpuapp]***: ***ROM*** size  ***increased by 4672[B]*** in comparison to the ***main[b02dd1b4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=applications&var-submodule=nrf_desktop&var-scenario=zrelease&var-testcase=none&var-arch=arm&var-platform=nrf54h20dk/nrf54h20/cpuapp) (cc: @MarekPieta)


*Note: This message is automatically posted and updated by the CI (latest/sdk-nrf/PR-17584/8)*

nordicjm commented on sysbuild/CMakeLists.txt:
```diff
@@ -226,6 +226,19 @@ function(${SYSBUILD_CURRENT_MODULE_NAME}_pre_cmake)
+    # The NRF54LX goes with PSA crypto by default
+    if(SB_CONFIG_SOC_SERIES_NRF54LX AND SB_CONFIG_BOOT_SIGNATURE_TYPE_ED25519)
+      set_config_bool(mcuboot CONFIG_NRF_SECURITY y)
+      set_config_bool(mcuboot CONFIG_MBEDTLS n)
+      set_config_string(mcuboot CONFIG_MBEDTLS_CFG_FILE "nrf-config.h")
```
```suggestion
```

nvlsianpu committed:
sysbuild/mcuboot: ed25519 pure signature integration

Added integration of signing image with the pure signature:
- New SB_CONFIG_BOOT_SIGNATURE_TYPE_PURE switch for enabling pure
signature.
- enforced sha512 for ed25519 on nRF54l target
- requests pure signature from imgtool.py

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on cmake/sysbuild/image_signing.cmake:
```diff
@@ -137,6 +137,22 @@ function(zephyr_mcuboot_tasks)
     set(input ${APPLICATION_BINARY_DIR}/zephyr/${KERNEL_NAME})
   endif()
 
+  if(SB_CONFIG_SOC_SERIES_NRF54LX AND SB_CONFIG_BOOT_SIGNATURE_TYPE_ED25519)
```
You would need to force usage of imgtool in ncs for this, look above for compression to see where, because the SDK toolchain is imgtool 2.1.0 which does not have this feature

de-nordic replied:
OK, so I have been able to avoid adding too many Kconfig options allover the place, by searching the imgtool extras for things that require our specific tool.

nordicjm replied:
No, imgtool extra is for user control, required options should not be there, use Kconfigs

nordicjm commented on cmake/sysbuild/image_signing.cmake:
```diff
@@ -137,6 +137,22 @@ function(zephyr_mcuboot_tasks)
+  if(SB_CONFIG_SOC_SERIES_NRF54LX AND SB_CONFIG_BOOT_SIGNATURE_TYPE_ED25519)
+    # Enforce SHA512 on nrf54l with ED25519; no need to bring SHA256 since ED25519
+    # is using SHA512 internally anyway.
+    set(imgtool_sha "--sha 512")
+  else()
+    set(imgtool_sha)
+  endif()
+
+  if(SB_CONFIG_BOOT_SIGNATURE_TYPE_PURE)
+    set(imgtool_pure "--pure")
+  else()
+    set(imgtool_pure)
```
just append to imgtool_extra, do not create new variables

de-nordic replied:
Did exactly that, but in Zephyr part.

nordicjm commented on sysbuild/Kconfig.mcuboot:
```diff
@@ -147,6 +147,24 @@ config MCUBOOT_FPROTECT_ALLOW_COMBINED_REGIONS
+	  hash of an image.
+	  This option is currently only supported with ED25519 and configurations
+	  where both image slots are within internal SoC device storage.
+
```
```suggestion
```

nordicjm commented on cmake/sysbuild/image_signing.cmake:
```diff
@@ -57,7 +57,14 @@ function(zephyr_mcuboot_tasks)
-  if(IMGTOOL AND NOT CONFIG_MCUBOOT_COMPRESSED_IMAGE_SUPPORT_ENABLED)
+  # Similar "pure" siganture and SHA512 needs NCS version of the imgtool,
+  # but for these we just check extra parameters.
+  string(REGEX MATCH "--sha |--pure[ $]?" imgtool_extension_required "${CONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS}")
```
why are there not Kconfigs for these? These should not be put into extra imgtool args, they should be Kconfigs which force the options

nvlsianpu commented on cmake/sysbuild/image_signing.cmake:
```diff
@@ -57,7 +57,14 @@ function(zephyr_mcuboot_tasks)
-  if(IMGTOOL AND NOT CONFIG_MCUBOOT_COMPRESSED_IMAGE_SUPPORT_ENABLED)
+  if(CONFIG_MCUBOOT_COMPRESSED_IMAGE_SUPPORT_ENABLED OR
+     (CONFIG_SOC_SERIES_NRF54LX AND CONFIG_MCUBOOT_BOOTLOADER_SIGNATURE_TYPE_ED25519))
+    set(NCS_IMGTOOL_REQUIRED "")
```
not used anymore, same check is preformed a few lines bellow.

de-nordic committed:
tests: kmu: Switch from MCUboot Kconfigs to sysbuild

Select KMU via sysbuild.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>
